[1.8.0] Cross branch and platform Auto-update

GeorgeK1ng opened the pull request:
Update metadata is automatically generated by GitHub Actions in the [vcmi/vcmi-updates](https://github.com/vcmi/vcmi-updates) repository and JSON files (vcmi-stable.json, vcmi-beta.json, vcmi-develop.json) are generated by [generate_update.py](https://github.com/vcmi/vcmi-updates/blob/master/.github/generate_update.py)

Each JSON contains version, commit, build date, changelog and per-platform download URLs
All branches and platforms which have generated packages become automatically available in the launcher
The launcher selects the correct download URL based on runtime OS and CPU architecture

This now brings automated update notifications for clients, with completely reworked update dialog - **VCMI Updates Center**

1) Windows - support full auto-update 
2) Android - a) VCMI installed from Google Play opens google play for download new version
                    b) Manually installed APK prompts for selecting folder where to save updated APK, user needs manual install
3) macOS - download DMG to selected folder, open it using default handler
4) iOS - if user have previously installed TestFligth or have iOS 16 and newer, then open TestFlight. When VCMI enters App Store, can be easily installed on iOS 12 from store.
5) Linux AppImage - support full auto-update 

Additional changes
1) Disabled update pop-up during first launch
2) Added visible green text - "New version available"
3) When enabled "Check for testing versions", it's automaticaly found newest version from Beta / Develop on that tab
4) Added help button with additional info for each operating system

How it looks now on Android
![1](https://github.com/user-attachments/assets/20af667f-73e5-4adc-9508-ecead6a8fb45)
![2](https://github.com/user-attachments/assets/199a9d22-d676-4244-9e54-a08de12febda)
![3](https://github.com/user-attachments/assets/ef30bb43-3d7e-40cf-81f3-a6c638c9f4b2)

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Update NumericConstants.h

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Add files via upload

GeorgeK1ng committed:
Add mobile style to UpdateDialog

GeorgeK1ng committed:
Refactor UpdateDialog for code consistency and clarity

kambala-decapitator commented:
ios: doesn't seem convenient for end user - you'd still need to transfer ipa to computer (at least for altstore). And with Testflight distribution it's completely useless.

Macos: easier to distribute in an archive instead of dmg for unattended install.

GeorgeK1ng commented:
I installed IPA on iOS using AltStore and IPA was added to AltStore directly from iPad. No need to have IPA on PC at least with Windows 10 + iPad mini 2 with iOS 12.5.7

Not sure what you mean with MacOS? It downloads DMG in Its form from vcmi.eu

GeorgeK1ng commented:
![Screenshot_20251229_202354](https://github.com/user-attachments/assets/7d8e81b0-a2d6-4ea6-815d-08bf47cfad88)
![Screenshot_20251229_202402](https://github.com/user-attachments/assets/3facbb20-a4a3-4eaf-8001-8a7be4c294c4)
![Screenshot_20251229_202411](https://github.com/user-attachments/assets/4af60fd4-41c1-4a41-bacb-5f642170b931)
![Screenshot_20251229_204232](https://github.com/user-attachments/assets/93c3c78b-a59d-4d8c-b46a-9618eb9f74ff)

GeorgeK1ng committed:
Update settings.json

GeorgeK1ng committed:
Refactor update dialog title and commit handling

Updated window title and text to 'VCMI Updates Center'. Replaced hardcoded commit SHA with currentCommit variable.

GeorgeK1ng committed:
Update updatedialog_moc.cpp

GeorgeK1ng commented:
So Install button on Release tab should open AppStore VCMI / Google Play page to install. But when on Testing tab allow select where to save IPA / APK. Probably require adding one more warning and point users to Installation Guide on GitHub / vcmi.eu.

kambala-decapitator commented:
> But when on Testing tab allow select where to save IPA / APK

to make it clear: iOS TestFlight has its own update mechanism, this built-in functionality isn't needed at all. And that 0.1% of sideloading users will update on their own, we won't make dedicated builds that has this internal update for them.

GeorgeK1ng committed:
Add files via upload

IvanSavenko commented on launcher/helper.cpp:
```diff
@@ -102,17 +105,21 @@ QString getRealPath(QString path)
+		const QAndroidJniObject jDst = QAndroidJniObject::fromString(dstIsContent ? safeEncode(dst) : dst);
+		QAndroidJniObject::callStaticMethod<void>("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;Landroid/content/Context;)V", jSrc.object<jstring>(), jDst.object<jstring>(), QtAndroid::androidContext().object());
+		return QFileInfo(dst).exists();
+	}
```
These (and some other) changes are already merged as part of first launch improvements, right?

Can you merge/rebase this branch onto beta to clean up diff's?

GeorgeK1ng replied:
I have already tried to upload same file which is currently in beta, but seems it's still marked like changes.

EDIT: Done.

GeorgeK1ng committed:
Merge pull request #5 from GeorgeK1ng/test1

Rebase

GeorgeK1ng committed:
Add createFileInTree into FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng committed:
Update helper.cpp

GeorgeK1ng committed:
Update FileUtil.java

GeorgeK1ng commented:
Now Android supports fully downloading APK to selected folder, user only needs to execute APK manually outside VCMI to update.

GeorgeK1ng committed:
Code review and additional improvements

GeorgeK1ng committed:
Fix progress bar position

GeorgeK1ng committed:
Support opening links from Changelog

GeorgeK1ng committed:
Update updatedialog_moc.cpp

GeorgeK1ng committed:
Add testing versions warning

GeorgeK1ng committed:
Auto update fix (#17)

* Fix new version detection

* Update updatedialog_moc.h

* Update updatedialog_moc.h

* Fix popup

GeorgeK1ng committed:
Refactor compareCandidateBuilds function parameters

GeorgeK1ng committed:
Leftover removal

Remove redundant check for identical commit hashes.

GeorgeK1ng committed:
Change update status message wording

GeorgeK1ng committed:
Improve Windows silent setup logic

GeorgeK1ng committed:
Update updatedialog_moc.cpp

GeorgeK1ng committed:
Update updatedialog_moc.cpp

GeorgeK1ng committed:
Add Linux mapping and code template fo future

GeorgeK1ng committed:
Fix stylus / mouse support on mobile platforms

GeorgeK1ng committed:
Small cleanup

GeorgeK1ng committed:
Enable scroll by swiping in Changelog

GeorgeK1ng committed:
Update updatedialog_moc.cpp

Laserlicht commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1032 @@
+#include <QSettings>
+#include <QPainter>
+#include <QTimer>
+#include <QScreen>
```
```suggestion
#include <QScreen>
#include <QProcessEnvironment>
```

Laserlicht commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1032 @@
+	return "ios-ios";
+
+#elif defined(VCMI_UNIX)
+	const auto arch = QSysInfo::buildCpuArchitecture();
```
```suggestion
	const auto arch = QSysInfo::buildCpuArchitecture();

	bool isAppImage = !qgetenv("APPIMAGE").isEmpty();
	if(!isAppImage)
	{
		logGlobal->warn("Auto Update on Linux is currently supported only for AppImage builds. Detected non-AppImage environment.");
		return {};
	}

```

Laserlicht commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1032 @@
+//        if (QFile::remove(currentAppImage)) { 
+//            if (QFile::rename(newFilePath, currentAppImage)) {
+//                // application can restarted
+//                QProcess::startDetached(currentAppImage);
+//                QCoreApplication::quit();
+//            } else {
+//                qDebug() << "Failed to rename new AppImage.";
+//            }
+//        } else {
+//            qDebug() << "Failed to remove old AppImage. Check permissions.";
+//        }
+
```
```suggestion
		QString currentAppImage = qgetenv("APPIMAGE");

		if(QFile::remove(currentAppImage))
		{ 
			if(QFile::rename(fullPath, currentAppImage))
			{
				// application can restarted
				QProcess::startDetached(currentAppImage);
				QCoreApplication::quit();
			}
			else
			{
				if(progress)
					progress->setVisible(false);

				ui->downloadLink->setText(tr("Failed to rename new AppImage."));
			}
		}
		else
		{
            if(progress)
				progress->setVisible(false);

            ui->downloadLink->setText(tr("Failed to remove old AppImage. Check permissions."));
		}
```

Laserlicht reviewed (commented):
With these changes auto update with appimage works. I've tested it.

https://github.com/Laserlicht/vcmi/commit/73082b509afd7a49405ad73ece8c615d97e26f28

GeorgeK1ng committed:
Apply suggestions for Linux AppImage update

Co-authored-by: Laserlicht <13953785+Laserlicht@users.noreply.github.com>

Laserlicht commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1025 @@
+#elif defined(VCMI_IOS)
+        // iOS: copy into the picked filesystem folder
+        {
+            const QString dstPath = QDir(target).filePath(fileName);
+            Helper::performNativeCopy(fullPath, dstPath);
+            Helper::revealDirectoryInFileBrowser(target);
+            ui->downloadLink->setText(tr("Saved to: %1 — install it manually.").arg(target));
+        }
+		if(progress)
+			progress->setVisible(false);
+
+#elif defined(VCMI_UNIX)
+		QString currentAppImage = qgetenv("APPIMAGE");
+
+		if(QFile::remove(currentAppImage))
+		{ 
+			if(QFile::rename(fullPath, currentAppImage))
+			{
+				// application can restarted
+				QProcess::startDetached(currentAppImage);
+				QCoreApplication::quit();
+			}
+			else
```
```suggestion
		QString currentAppImage = qgetenv("APPIMAGE");
		
		if(currentAppImage.isEmpty())
		{
			if(progress)
				progress->setVisible(false);

			ui->downloadLink->setText(tr("No AppImage found."));
			return;
		}

		if(QFile::remove(currentAppImage))
		{ 
			if(QFile::rename(fullPath, currentAppImage))
			{
				// application can restarted
				QProcess::startDetached(currentAppImage, QStringList());
				QCoreApplication::quit();
			}
			else
```

Laserlicht reviewed (changes requested):
Sorry - again. Test failed because `startDetached` signature deprecated...

Also added `currentAppImage.isEmpty()` check. Looks better in UI.

GeorgeK1ng committed:
Add TestFlight for iOS >= 16

GeorgeK1ng committed:
Fix tests

GeorgeK1ng committed:
Apply suggestions from code review

Co-authored-by: Laserlicht <13953785+Laserlicht@users.noreply.github.com>

kambala-decapitator commented:
Testflight itself works on iOS >= 12, works on my iPhone 6+ for instance. But if you never had TF app installed (not "purchased"), then you can do so only on iOS >+ 16 at the moment.

To sum up: if a user is already using TF, then you don't need any condition checks. But you need TF detection.

GeorgeK1ng committed:
Add TestFlight check

GeorgeK1ng commented:
Both iOS and Linux cases should be covered now.

IvanSavenko commented on launcher/helper.h:
```diff
@@ -21,11 +21,13 @@ namespace Helper
 	void nativeFolderPicker(QWidget *parent, std::function<void(QString)>&& cb);
 	QStringList findFilesForCopy(const QString &treeUri);
 	void sendFileToApp(QString path);
+	bool isInstalledFromGooglePlay();
```
```suggestion
#ifdef VCMI_ANDROID
	bool isInstalledFromGooglePlay();
#endif
```

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
 {
 	ui->setupUi(this);
-	
+	ui->buildChannel->setItemData(0, "develop");
+	ui->buildChannel->setItemData(1, "beta");
```
Can you explain why we have this strange split between "stable" and "testing" channels, with "testing" having effectively own "subchannels" for "beta" and "develop"?

Why not simply have 3 channels:
- stable
- beta
- develop

?

GeorgeK1ng replied:
Stable - simply enabled for everyone. Beta / Develop branches are considered as testing builds. That's why they are splitted. Don't see reason to show casual players dropdown with 2 more options.

IvanSavenko replied:
But we still show them stable / testing toggle. Don't see much difference between showing stable/testing selection and stable/beta/unstable selection

GeorgeK1ng replied:
Yes, but thats on another tab, where you need to enable "Testing checkbox". If you not enable them, you will never notice other builds.

IvanSavenko replied:
So what's the difference in UX between showing "Testing checkbox" and showing "Testing combobox"?

In both cases you have UI element to switch between stable / unstable, but in your case you have two of them for some reason - checkbox AND combobox (on another tab)

IvanSavenko replied:
Looking on your UI, you have 3 (!) elements to indicate channel:
- Tabs
- Checkbox on testing tab
- Dropdown on testing tabs

So:

1. Why we even need tabs?
2. Why we need that checkbox?

Just show "Updates channel" dropdown, that contains Stable / Beta / Unstable selector

GeorgeK1ng replied:
You can't click on anything on testing tab until you enable it - then you reveive unstable warning popop. Then you can select which testing branch you want to see or download.

IvanSavenko replied:
We can show that warning on selecting beta / develop in dropdown

GeorgeK1ng replied:
Why should casual user see other branches? One comboxo with 3 options easy to mess with mouse wheel and download / install unwated version. Same on mobiles, easy to mess combobox selection.

Better question, is what's wrong in current UI design / logic?

IvanSavenko replied:
> Better question, is what's wrong in current UI design / logic?

I don't see any point in separating "do I want testing builds" from "which testing build do I want".

For example, right now if I want to install develop build I have to enable testing checkbox and then in dropdown select develop build. What's the point of having such two steps? 

What's the point of having 2 tabs that display basically same info but for different builds? Which version will I get if I have testing enabled, but release tab active and press Update button?

The way I see it, we can replace "enable testing' checkbox with "select update channel". If your concern is accidental mouse scroll then we can instead use 3 toggle buttons to switch between channels, like in settings menu

GeorgeK1ng replied:
I also want to prevent accidentaly joined beta / develop versions. 

Download button currently works on what you see, if you are on testing tab and develop branch it will download it.

Also each system have different behavior, On Windows all 3 can coexist, on Android release and one of testing version as daily. 

And separated tab is for me fine to not catch too much testers when turning to new version with new mod format, which can crash game etc.. Hope testing tab will be used by VCMI moders and active comunity to help report and fix issues.

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
+	QMessageBox::information(this, tr("Update channels information"), details + "\n\n" + common);
+}
+
+void UpdateDialog::on_installButton_clicked()
```
Please split this massive method into logical parts, or extract some code into a separate method

GeorgeK1ng replied:
Done

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
+#elif defined(VCMI_IOS)
+	return QObject::tr("You are running iOS. Usually only one VCMI app installation is active at a time, and installing another build typically replaces the previous app.");
+#elif defined(VCMI_UNIX)
+	return QObject::tr("You are running Linux. Updates are currently supported only for AppImage builds.");
```
```suggestion
	return QObject::tr("You are running Linux. Updates are currently supported only for AppImage builds. Please use your system package manager to update VCMI");
```

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
+static QString normalizeChannel(const QString& text)
+{
+	const auto str = text.trimmed().toLower();
+	if(str.contains("beta"))
+		return "beta";
```
What happens once we release last beta build & after some time - remove all beta builds? 

We need some logic for "switch from beta to release if release is newer"

GeorgeK1ng replied:
Depends on branch, is you are on Beta branch and have testing enabled a no download available - should turn into message, no download available for your platform. These cases are platform specific.

For example on Android it will be VCMI Daily and latest build from beta / develop should be offered. User still can download and install stable on from stable tab, but can't check if it's really newer that installed VCMI. As it's like download manager only in that case. 

And I hope I have covered these scenarios, but don't remember exactly all tested scenarios.

IvanSavenko replied:
Do we have detection for "no download available"? Our server currently configured to auto-delete all files older than 30 or 45 days. What would happen to players who pick beta channel after all beta builds were removed due to age?

GeorgeK1ng replied:
Should be fine too with message about no downloads available.

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
+	logText += "<br/><br/>"; // blank line between header and body
+	logText += body;
+
+	if(changelogBox)
+		changelogBox->setMarkdown(logText);
```
How changelog works? Uses Changelog.md for stable? What about develop / beta?

GeorgeK1ng replied:
Beta / Develop have currently static message. Stable changelog is text from release - https://github.com/vcmi/vcmi/releases into stable.json

IvanSavenko commented on launcher/updatedialog_moc.cpp:
```diff
@@ -13,135 +13,1062 @@
+	version = QString::fromStdString(newVersion);
+
+	if(link.isEmpty())
+		changelogBox->setMarkdown(tr("No download available for this platform."));
```
In which cases this might happen?
Google Play? Non-AppImage Linux?

GeorgeK1ng replied:
Android x86, Portmaster or any other build which is not in json update file

IvanSavenko replied:
Then perhaps replace this string with something like "Automatic updates are not supported on your platform. Please install updated version manually"? To actually tell player what he needs to do when this happens

GeorgeK1ng replied:
Done with "Automatic updates are not supported for this platform/channel combination. Please install the updated version manually."

GeorgeK1ng committed:
Update launcher/updatedialog_moc.cpp

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>

GeorgeK1ng committed:
Update launcher/helper.h

Co-authored-by: Ivan Savenko <saven.ivan@gmail.com>

GeorgeK1ng committed:
Add native file picker when native folder picker not available

GeorgeK1ng committed:
Review adjustments

GeorgeK1ng committed:
Update updatedialog_moc.cpp

GeorgeK1ng committed:
ReRun CI

GeorgeK1ng commented:
With new changelog generated from PRs names tooks testing tab really good. Also PR numbers are clickable, so we can easily check whole PR if needed.

<img width="549" height="338" alt="image" src="https://github.com/user-attachments/assets/1a6cc807-a5c4-4408-82e9-5e1bc04e8361" />

GeorgeK1ng committed:
Revert CanUseFolderPicker solution

GeorgeK1ng committed:
ReRun CI

GeorgeK1ng commented:
Can we please go to merge with this one? Want to have some time to tweak it.

IvanSavenko commented:
Tweak this PR then to make it more manageable. Currently it is constantly in the back of my "to review" list because reviewing 1500 lines of extremely complex, heavily branching and poorly written code is massive pain.

For example, I still don't have clear idea how it works on every platform and on every branch since this logic is different everywhere. What's worse - this logic is spread in a thin layer across several 200+ line methods, half of which are 'if' checks.

It also does A LOT of things:
- new update dialog
- support for multiple channels
- support for auto-update on multiple different platforms, each with own logic
- as well as a lot of supporting code for such functionality

Honestly splitting it into several parts would make it way easier to review. For example as a starting point, we could only provide new update checking dialog for stable builds, and add multiple branches in separate PR, then - add auto-update for one platform at a time.

GeorgeK1ng commented:
To be honest, this feedback is extremely discouraging.

I started working on this in September 2025 and opened the PR in December 2025. Since then, I've spent a lot of time implementing requested changes, fixing issues, and adding the supporting infrastructure needed to make the feature actually work.

Now, after all this time, the main feedback is that the entire feature should be split into multiple PRs. If that had been the expectation from the beginning, I would have approached the implementation differently.

At this point, I don't have the motivation to restructure and effectively reimplement tons of work based on a requirement that was never raised earlier. The PR is already working, including the update dialog, channel support, Windows auto-update functionality, and the update metadata generation work in a separate repository.

The PR has already started accumulating merge conflicts due to the time it has spent waiting for review, and that was the final straw for me.

If splitting this work into several separate PRs is a requirement for merging, I'll leave the PR as-is and won't be spending any more time on it. If someone else wants to continue from here, they are welcome to use and adapt the code I've already contributed.

IvanSavenko commented:
Yeah, should've mentioned that sooner. This is often problem with your PR's - you often make large PR's with multiple complex changes that can be easily split into parts and done in stages. 

Because of that it is quite hard for me to review it, so I would often try to do it in stages, posponing most complex parts till I get a good picture of what the code actually does.

If you wish more concrete points (that would have been easy to discover and fix over time in case of modular PR) then:
- Code is hard to manage. In its current state I am not sure if we should merge it because any bug would be hard to track down. 
- updatedialog_moc is tangled mess of free functions with numerous ifdef's and if's. Creating either interface or class with multiple separate implementations depending on platforms would have made it more manageable
- UpdateDialog class itself mixes UI handling and actual business logic. Sure, common in our Launcher, but in this case logic is complex as it is and any bug is VERY dangerous since it can do things like file removal. And I remember there was bug around or shortly before 1.0 where mod removal would wipe entire home directory.
- And I still don't understand why we need this complex UX with multi-staged selection of desired channel

I am sure this is not a complete list. For example I looked through your PR right now and noticed one more scenario - from what I see, builds from PR would attempt to upgrade to latest develop build, which is not expected behavior. And I am sure that there is more, but it is hard to find such cases in large tangled code.

IvanSavenko commented:
Might try to recover this PR later, cleaning up the code along the way, but no ETA / no promises for now.

GeorgeK1ng commented:
If the user enables "testing builds", then yes, update notifications will be offered. However, on Windows, stable, beta, and develop builds coexist, and users are clearly warned when opting into unstable builds.

Updates are also branch-aware. For example, if you're running a develop build, develop updates are prioritized and offered first.

Yes, there are a lot of platform-specific conditions in the code, but that's intentional and helps avoid duplicating update logic across different platforms and packaging formats.

One positive aspect is that there are no issues with file replacement/removal on Windows. The Linux AppImage implementation was not written by me, but it has been tested as well. On Windows, the setup is quite straightforward and handled by the installer.

The reason for allowing users to choose which branch to follow, rather than always forcing the latest available version, is to support testers and modders. We could potentially add another checkbox in the Launcher settings for this, but I don't think it's really necessary.

From my perspective, this was already in a state where it could have been merged and polished further over the following months. There is simply no better way to test a feature like this than by having it used in the real world.
